### DISCUSSION TITLE
Coupling Velocity with other variables

### DISCUSSION THREAD

**Discussion by alimostafavi24** [idaholab/moose#23254] [Q&A Getting Started] [answered]
Hello everyone, 

I am new to MOOSE and trying to solve coupled PDE equation (Cahn-Hilliard Navier-Stokes). My problem is that I do not how to correctly couple velocity and concentration. 

This problem is similar to the convection heat transfer when we need to couple the velocity and temperature.

It would be much appreciated if someone can guide me here by introducing a relevant input file or explaining the way that the kernels must be set.

Regards,
Ali

**Comment by GiudGiud** [comment]
Hello

Are you using the finite element or finite volume implementations of Navier Stokes?
For finite volume, you ll want to have a look at this input that transports a scalar concentration using the determined velocity
https://github.com/idaholab/moose/tree/next/modules/navier_stokes/test/tests/finite_volume/ins/multiapp-scalar-transport

Guillaume

**Comment by alimostafavi24** [comment]
Hi,

Thank you for your answer!

I am actually using the finite element method to couple Navier stokes and Cahn hilliard equation which requires velocity coupling.

Thanks,
Ali

**Reply by GiudGiud** [reply]
Hello

we don’t have scalar transport in finite element navier stokes right now. 

You can implement it fairly easily from the energy advection kernels.

Guillaume

**Comment by alimostafavi24** [comment]
I am trying to couple these 2 PDEs. So, both c and u are coupled variables.

[image: Screen Shot 2023-01-27 at 12.25.36 PM.png]
[image: Screen Shot 2023-01-27 at 12.26.02 PM.png]
[image: Screen Shot 2023-01-27 at 12.26.10 PM.png]

My question is that are there specific kernels that I can use to couple
those variables or I should develop my own kernels to couple them?

Thanks,
Ali

**Reply by GiudGiud** [reply]
The images did not work. 

If you are doing scalar transport with finite element in more than one dimension, then you need your own kernels. 
Unless you have access to Griffin ? (it s an export controlled reactor physics code)

**Reply by alimostafavi24** [reply]
![image](https://user-images.githubusercontent.com/119982603/215183974-58114106-bbb0-417c-ad27-e60a085d85bc.png)
![image](https://user-images.githubusercontent.com/119982603/215184062-0cb5bfcc-a7e3-4059-be46-ffed8d557ee1.png)
![image](https://user-images.githubusercontent.com/119982603/215184101-5d2290b1-51de-460a-a723-37cd2e033e36.png)
  
I think now equations can be seen. No, I don't have access to Griffin. And I want to solve it in 2D.

**Reply by GiudGiud** [reply]
ok then you need to implement a set of kernels, and we would be interested in the contribution of these afterwards btw

To do that you may:
- have a look at energy advection FE kernels. Energy and scalars are advected similarly
- have a look at the scalar advection FV kernels, consider verifying your work with a similar FV solution

**Reply by alimostafavi24** [reply]
All right! I had one more question. I developed a custom kernel myself. Is there any way that during compilation this can be added to combined-opt? For example, I added the .h and .c file of my custom kernel to Navier-stokes module and it did not work. What is the best way to add our custom kernels to existing MOOSE kernels and compile everything?

**Reply by GiudGiud** [reply]
it should be added to the combined-opt if:
- you placed the files in navier_stokes/src/kernels and navier_stokes/include/kernels respectively
- you recompiled the navier stokes module or combined afte
- they were registered to the right application

**Reply by alimostafavi24** [reply]
It is odd! I write my custom kernel (AdvectionC) and put in a correct path on navier-stokes and phase-field modules and compiled them. Then I compiled for combined-opt and use it to run my input file. This is the error I got.

A 'AdvectionC' is not a registered object.

If you are trying to find this object in a dynamically linked library, make sure that
the library can be found either in your "Problem/library_path" parameter or in the
MOOSE_LIBRARY_PATH environment variable.

**Reply by GiudGiud** [reply]
What name did you use for registering?

it needs to be exact

**Reply by alimostafavi24** [reply]
It is. This is from the input file.

  [./conv]
    type = AdvectionC
     variable = c
	velocity_x = vel_x
	velocity_y = vel_y
 [../]

**Reply by GiudGiud** [reply]
Registering is in the source file.
You should have something like:
```
registerMooseObject("NavierStokesApp", PressureGradient);
```

**Reply by alimostafavi24** [reply]
I understand the problem. Now it is working fine.

Thanks,
Ali

**Reply by GiudGiud** [reply]
AD or non AD is independent of vector or by-component for velocities

You can retrieve velocities by component using the coupledValue() routine in the constructor
or adCoupledValue('vel_x') for example if you want to use AD

**Reply by alimostafavi24** [reply]
Could you please explain more about adCoupledValue('vel_x') (For example, I want to know where I should implement it.)?

I also have the very same problem for material property. In my input file, I defined rho and mu as regular material property, but when I run it I see the following error:

`The requested AD material property mu is declared as a regular material property. Either retrieve it as a regular material property with getMaterialProperty or declare it as an AD property with declareADProperty`

How can I resolve this problem?

Thanks,
Ali

**Reply by GiudGiud** [reply]
You need to define rho and mu as AD properties. For example, instead of this:
```
[Materials]
  [./k]
    type = GenericConstantMaterial
    prop_names = 'thermal_conductivity'
    prop_values = '10' # in W/mK
  [../]
[]
```
you need to use the AD version:
```
[Materials]
  [./k]
    type = ADGenericConstantMaterial
    prop_names = 'thermal_conductivity'
    prop_values = '10' # in W/mK
  [../]
[]
```

You can see examples of using adCoupledValue in the ADMatReaction kernel

**Reply by alimostafavi24** [reply]
I see. But for me rho is not constant as it is function of c. I am using DerivativeParsedMaterial to define rho. I also cannot use AD because I am using some other kernels. I think I should change INSADStabilized3Eqn in a way that it does not require ADMaterialProperty and accept DerivativeParsedMaterial.

**Reply by GiudGiud** [reply]
You can mix AD and regular kernels. However you need to make sure that every kernel receives the material property that they want, with the AD/not AD type wanted

**Reply by alimostafavi24** [reply]
Can you show me how can I define rho as a function of c as a mixed AD and regular kernel?

**Reply by GiudGiud** [reply]
rho is a material property right? For example:
```
[material_1]
  type = ParsedMaterial
  property_name = 'rho'
  coupled_variables = 'c'
  expression = '1 + c + c*c'
[../]
[material_2]
  type = ADParsedMaterial
  property_name = 'rho_ad'
  coupled_variables = 'c'
  expression = '1 + c + c*c'
[../]
```

then you use rho and rho_ad for the kernels that expect one or the other.
A note of caution, for every non AD kernel you will likely need to take into account the derivative of rho with regards to c

**Comment by alimostafavi24** [comment]
Hello,

I think using ADEnergyAdvection is problematic for my problem. I need another kernel to couple velocity and concentration that does not require velocity as a vector or AD kernel. Do you have any suggestion that might help me?

Thanks,
Ali

**Reply by GiudGiud** [reply]
You can try this one
https://mooseframework.inl.gov/source/kernels/ConservativeAdvection.html

but at the end of the day, an advection kernel is not the hardest thing to make so you could also just code one

**Reply by alimostafavi24** [reply]
Hello,

I have one more question!

```
[Materials]
  [./const]
    type = ADGenericConstantMaterial
    prop_names = 'rho mu cp k'
    prop_values = '1  1  1  .01'
  [../]
  [ins_mat]
    type = INSADStabilized3Eqn
    velocity = velocity
    pressure = p
    temperature = temperature
  []
[]
```
I want to use ADGenericFunctionMaterial. Do you know how I can describe mu and rho as a function of c (They are not constant in my problem)? I tried different ways and all gave me an error.

Thanks,
Ali

**Reply by GiudGiud** [reply]
You can use a ParsedMaterial for that instead. Functions are meant to be pre-determined in space and time at the beginning of the simulation, they cant have local variation on local quantities

**Reply by alimostafavi24** [reply]
Ok. I will use ADParsedMaterial. What about the syntax (considering the simple case that rho='c' and mu='c')?

**Reply by GiudGiud** [reply]
I pasted ParsedMaterial examples above a few days ago. The syntax should be fairly clear from those

**Reply by alimostafavi24** [reply]
Yes, it is. But I cannot use that. My input file is similar to this (modules/navier_stokes/test/tests/finite_element/ins/RZ_cone/ad_rz_cone_no_parts_steady_stabilized.i). When I used that kind of syntax which you mentioned previously, I get this error.
```

'Materials/const/type'
	Doc String: "A string representing the Moose Object that will be built by this Action"
```

Is there anyway that we can modify the below part and express rho and mu as a function of c?

```
  [./const]
    type = ADGenericConstantMaterial
    prop_names = 'rho mu cp k'
    prop_values = '1  1  1  .01'
  [../]

```

**Reply by GiudGiud** [reply]
no there is not. A constant cannot become a parsed expression in this context
What you pasted is not an error either or maybe you forgot one part of it.

**Reply by alimostafavi24** [reply]
Right. Regarding developing a custom advection kernel, I will start with the ConservativeAdvection which requires velocity as a vector. Could you please the main things that I should look for in order to modify this using velocity components?

Thanks,
Ali

**Reply by GiudGiud** [reply, marked as answer]
that part should be easy:
coupledVectorValue -> coupledValue in the constructor
VectorVariableValue -> VariableValue in the header
and add the other components

the to access them: _vel_x[_qp] and same for yz

**Reply by alimostafavi24** [reply]
Right. This is the error that I get after substituting those in the constructor.

```
/home/ali/projects/moose/modules/navier_stokes/src/kernels/ConservativeAdvectionn.C: In member function 'libMesh::Real ConservativeAdvectionn::negSpeedQp() const':
/home/ali/projects/moose/modules/navier_stokes/src/kernels/ConservativeAdvectionn.C:49:31: error: cannot convert 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<double> >::type' {aka 'libMesh::TypeVector<double>'} to 'libMesh::Real' {aka 'double'} in return
   49 |   return -_grad_test[_i][_qp] * _velocity_x[_qp];
/home/ali/projects/moose/modules/navier_stokes/src/kernels/ConservativeAdvectionn.C:50:31: error: cannot convert 'libMesh::boostcopy::enable_if_c<true, libMesh::TypeVector<double> >::type' {aka 'libMesh::TypeVector<double>'} to 'libMesh::Real' {aka 'double'} in return
   50 |   return -_grad_test[_i][_qp] * _velocity_y[_qp];
make: *** [/home/ali/projects/moose/framework/build.mk:145: /home/ali/projects/moose/modules/navier_stokes/build/unity_src/kernels_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
```
